test/utils/dirft2d.hpp

```diff
@@ -43,63 +58,82 @@ c     used, otherwise the - sign is used, in the exponential.
   }
 }
 
-template<typename T>
-void dirft2d2(BIGINT nj, T *x, T *y, std::complex<T> *c, int iflag, BIGINT ms, BIGINT mt,
-              std::complex<T> *f)
-/* Direct computation of 2D type-2 nonuniform FFT. Interface same as finufft2d2
+// ------------------------------------------------------------
+// 2D Type-2 NUFFT, direct:
+//   c[j] = Σ_{k1,k2} f[k1,k2] exp(i * iflag * (k1 x[j] + k2 y[j]))
+// ------------------------------------------------------------
+template<typename BIGINT, typename XYArr, typename CArr, typename FArr>
+void dirft2d2(BIGINT nj,
+              const XYArr &x,
+              const XYArr &y,
+              CArr &c,
+              int iflag,
+              BIGINT ms,
+              BIGINT mt,
+              const FArr &f) {
+  using Complex = std::decay_t<decltype(c[0])>;
+  using FLT     = typename Complex::value_type;
+
+  static_assert(std::is_floating_point<FLT>::value,
+                "Complex<FLT> must be floating point");
+  static_assert(std::is_same<std::decay_t<decltype(x[0])>, FLT>::value,
+                "x[j] must be FLT");
+  static_assert(std::is_same<std::decay_t<decltype(y[0])>, FLT>::value,
+                "y[j] must be FLT");
 
-     c[j] = SUM   f[k1,k2] exp(+-i (k1 x[j] + k2 y[j]))
-            k1,k2
-                            for j = 0,...,nj-1
-    where sum is over -ms/2 <= k1 <= (ms-1)/2,  -mt/2 <= k2 <= (mt-1)/2.
+  const BIGINT k1min = -(ms / 2), k2min = -(mt / 2);
+  const Complex I0{0, FLT(iflag)};
 
-    The input array is in increasing k1 ordering (fast), then increasing
-    k2 ordering (slow).
-    If iflag>0 the + sign is used, otherwise the - sign is used, in the
-    exponential.
-    Uses C++ complex type and winding trick.  Barnett 1/26/17
-*/
-{
-  BIGINT k1min = -(ms / 2), k2min = -(mt / 2); // integer divide
   for (BIGINT j = 0; j < nj; ++j) {
-    std::complex<T> a1  = (iflag > 0) ? exp(std::complex<T>(0, 1) * x[j])
-                                      : exp(-std::complex<T>(0, 1) * x[j]);
-    std::complex<T> a2  = (iflag > 0) ? exp(std::complex<T>(0, 1) * y[j])
-                                      : exp(-std::complex<T>(0, 1) * y[j]);
-    std::complex<T> sp1 = pow(a1, (T)k1min);
-    std::complex<T> p2  = pow(a2, (T)k2min);
-    std::complex<T> cc  = std::complex<T>(0, 0);
-    BIGINT m            = 0; // input pointer
+    Complex a1  = exp(I0 * x[j]);
+    Complex a2  = exp(I0 * y[j]);
+    Complex sp1 = pow(a1, FLT(k1min));
+    Complex p2  = pow(a2, FLT(k2min));
+    Complex acc{0, 0};
```

Comment:
the name "cc" was for consistency with the other routines... a component of c.  (not "acc" for accumulation).